src/vsr/replica.zig

```diff
@@ -2402,13 +2402,48 @@ pub fn ReplicaType(
             self.sync_reclaim_tables();
         }
 
+        fn on_request_sync_checkpoint(self: *Self, message: *Message) void {
+            assert(message.header.command == .request_sync_checkpoint);
+            if (self.ignore_sync_request_message(message)) return;
+
+            assert(message.header.op ==
+                self.superblock.staging.vsr_state.checkpoint.commit_min);
+            assert(message.header.parent == self.superblock.staging.checkpoint_id());
+
+            self.send_sync_checkpoint(.{ .replica = message.header.replica });
+        }
+
+        fn on_sync_checkpoint(self: *Self, message: *const Message) void {
+            assert(message.header.replica < self.replica_count);
+            assert(message.header.command == .sync_checkpoint);
+            if (self.ignore_sync_response_message(message)) return;
+
+            const stage: *SyncStage.RequestingTrailers = &self.syncing.requesting_trailers;
+            assert(stage.target.checkpoint_id == message.header.parent);
+
+            log.debug("{}: on_sync_checkpoint: checkpoint_op={} checkpoint_id={x:0>32}", .{
+                self.replica,
+                stage.target.checkpoint_op,
+                stage.target.checkpoint_id,
+            });
+
+            const checkpoint_state = std.mem.bytesAsValue(
+                vsr.CheckpointState,
+                message.body()[0..@sizeOf(vsr.CheckpointState)],
+            );
+
+            if (stage.done()) {
+                self.sync_requesting_trailers_callback(checkpoint_state);
+            }
```

Comment:
When the client sessions trailer is removed, the client sessions will move into the `sync_checkpoint` message body too, immediately following the `CheckpointState` in the message body.